### ISSUE TITLE
Add Swift Package Manager (SPM) file support

### ISSUE THREAD

**Issue by matis-schotte** [ResearchKit/ResearchKit#1111]
Lets add support for SPM via a `Package.swift` file, next to the already existing CocoaPods support.

**Comment by bgannin** [comment]
I'll take a look at this tonight or over the weekend.

**Comment by berenyiaustin** [comment]
@bgannin do you know if any steps were taken with this? I would love SPM support for RK.

**Comment by damienpontifex** [comment]
I had a brief try at this today. Hit a issue early on that there's mixed language used in the project. So get the error `target ResearchKit contains mixed language source files; feature not supported`.

Maybe someone with more experience knows how to create Package.swift with multiple targets or something to resolve this? Otherwise an effort to migrate to a singular language?

**Comment by igoreydman** [comment]
Any updates to this issue? I tried importing ReasearchKit through SPM and hit an error that there is no Package.swift manifest.

<img width="520" alt="Screen Shot 2020-07-08 at 10 01 32 AM" src="https://user-images.githubusercontent.com/10006991/86928008-023f0000-c102-11ea-81a1-18e234c0cdcc.png">

**Comment by tmbiOS** [comment]
> Any updates to this issue? I tried importing ReasearchKit through SPM and hit an error that there is no Package.swift manifest.
> 
> <img alt="Screen Shot 2020-07-08 at 10 01 32 AM" width="520" src="https://user-images.githubusercontent.com/10006991/86928008-023f0000-c102-11ea-81a1-18e234c0cdcc.png">

Any updates to this issue?

**Comment by smileypop** [comment]
For an open source framework that is unofficially "sponsored" by Apple, it's crazy that SPM support hasn't been introduced yet :(

**Comment by SwiftNativeDeveloper** [comment]
What would we expect to be provided as libraries? ResearchKit + ResearchKitUI?

Taking a play from many of Apple SDKs in the OS, as well as sibling library [CareKit](https://github.com/carekit-apple/CareKit/blob/main/Package.swift) I can imagine there would be some natural splits which could be a bit of work to refactor/reorganize this source code.

To socialize the idea, for those who are more familiar with the code structure, are there other components that stand out as their own library component of a future ResearchKit SPM?

**Comment by jscubida** [comment]
ResearchKit isn't structured like CareKit, so it would require considerable refactoring to split it into 2 packages. Furthermore, it's still written in Objective-C, so that's another hurdle.

**Comment by SwiftNativeDeveloper** [comment]
@jscubida yep, "bit of work" lacked emphasis above :)

**Comment by j-dogcoder** [comment]
Hey, is there any progress on this? Would love to see this supported, as SPM is a very useful tool.

**Comment by j-dogcoder** [comment]
Would love to see this @Pariecemckinney-apple!

**Comment by PSchmiedmayer** [comment]
Hi everyone,

I just wanted to share that we created a fork of the ResearchKit project over at https://github.com/StanfordBDHG/ResearchKit that supports the Swift Package Manager. 

You can import it into your project or add it to your Swift Package by following the article about [Adding Package Dependencies to Your App](https://developer.apple.com/documentation/xcode/adding-package-dependencies-to-your-app) using the following repository URL: `https://github.com/StanfordBDHG/ResearchKit`. 

While this is not an official SPM support by Apple, it might help you with your projects.

**Comment by SwiftNativeDeveloper** [comment]
> Hi everyone,
> 
> I just wanted to share that we created a fork of the ResearchKit project over at https://github.com/StanfordBDHG/ResearchKit that supports the Swift Package Manager.
> 
> You can import it into your project or add it to your Swift Package by following the article about [Adding Package Dependencies to Your App](https://developer.apple.com/documentation/xcode/adding-package-dependencies-to-your-app) using the following repository URL: `https://github.com/StanfordBDHG/ResearchKit`.
> 
> While this is not an official SPM support by Apple, it might help you with your projects.

Hey @PSchmiedmayer, I added your SPM as a reference and uploaded a build to TestFlight without any imports to ResearchKit. This upload produced ITMS-90683: Missing Purpose string in Info.plist for the following types:
- NSMicrophoneUsageDescription
- NSCameraUsageDescription
- NSMotionUsageDescription
- NSSpeechRecognitionUsageDescription
- NSLocationAlwaysAndWhenInUseUsageDescription
- NSLocationWhenInUseUsageDescription
- NSHealthShareUsageDescription
- NSHealthUpdateUsageDescription
- NSHealthClinicalHealthRecordsShareUsageDescription

I'm not sure if I can change the compile settings to more aggressively strip the unused parts of ResearchKit (like motion, location, etc.) or not to avoid this ITMS error. My suspicion is because in your case, ResearchKit is a binary XCFramework and it can't dead code strip enough out. At least not without being told to do so, still looking into that bit.

What are you doing to overcome these warnings? I assume just providing a usage description string and either:
1. Make use of said permission string
2. Never invoke the API to trigger the capability

**Comment by SwiftNativeDeveloper** [comment]
Re: https://github.com/ResearchKit/ResearchKit/issues/1111#issuecomment-996909877
> What would we expect to be provided as libraries? ResearchKit + ResearchKitUI?
> 
> Taking a play from many of Apple SDKs in the OS, as well as sibling library [CareKit](https://github.com/carekit-apple/CareKit/blob/main/Package.swift) I can imagine there would be some natural splits which could be a bit of work to refactor/reorganize this source code.
> 
> To socialize the idea, for those who are more familiar with the code structure, are there other components that stand out as their own library component of a future ResearchKit SPM?

and https://github.com/ResearchKit/ResearchKit/issues/1111#issuecomment-996926418
> ResearchKit isn't structured like CareKit, so it would require considerable refactoring to split it into 2 packages. Furthermore, it's still written in Objective-C, so that's another hurdle.

Well now, looks like ResearchKit got a modularization update!

https://github.com/ResearchKit/ResearchKit/releases/tag/3.0

What else needs to happen before a Swift Package can be added by Apple for this repo? 😄

**Comment by Pariecemckinney-apple** [comment]
Hello @SwiftNativeDeveloper,

Thanks for reaching out. SPM support has been at the top of our minds for quite some time. However, getting around the lack of [mixed language target support](https://github.com/swiftlang/swift-evolution/blob/main/proposals/0403-swiftpm-mixed-language-targets.md) has created a significant hurdle at the moment.

Another option we could explore in the meantime would be a binary SPM package. Would that be of interest to you?

We’ll continue to explore options on our side and make sure to ping folks here once we have an update ready.

**Comment by PSchmiedmayer** [comment]
@SwiftNativeDeveloper If you are interested in a binary distribution path that @Pariecemckinney-apple mentioned, you can check out https://github.com/StanfordBDHG/ResearchKit; contributions there are always welcome.

**Comment by SwiftNativeDeveloper** [comment]
@PSchmiedmayer @Pariecemckinney-apple 

We are already pointing to the distribution by Stanford. It requires that we provide usage description strings for features that we do not use and I want to harden the app privacy.

This was mentioned a few years ago above. I assume this is still a prerequisite to using the library in that manner.

https://github.com/ResearchKit/ResearchKit/issues/1111#issuecomment-1437193888
